crates/ruff_linter/src/rules/ruff/rules/dummy_variable_accessed.rs

```diff
@@ -0,0 +1,246 @@
+///
+/// ## Example
+/// ```python
+/// def function():
+///     _variable = 3
+///     return _variable + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def function():
+///     variable = 3
+///     return variable + 1
+/// ```
+///
+/// ## Options
+/// - [`lint.dummy-variable-rgx`]
+#[derive(ViolationMetadata)]
+pub(crate) struct DummyVariableAccessed {
+    name: String,
+    fix_kind: Option<ShadowedKind>,
+}
+
+impl Violation for DummyVariableAccessed {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Local dummy variable `{}` is accessed", self.name)
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        if let Some(fix_kind) = self.fix_kind {
+            return Some(match fix_kind {
+                ShadowedKind::BuiltIn => {
+                    "Prefer using trailing underscores to avoid shadowing a built-in".to_string()
+                }
+                ShadowedKind::Keyword => {
+                    "Prefer using trailing underscores to avoid shadowing a keyword".to_string()
+                }
+                ShadowedKind::Some => {
+                    "Prefer using trailing underscores to avoid shadowing a variable".to_string()
+                }
+                ShadowedKind::None => "Remove leading underscores".to_string(),
+            });
+        }
+        None
+    }
+}
+
+/// RUF052
+pub(crate) fn dummy_variable_accessed(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    let name = binding.name(checker.source());
+
+    // Ignore `_` and dunder variables
+    if name == "_" || is_dunder(name) {
+        return None;
+    }
+    // only used variables
+    if binding.is_unused() {
+        return None;
+    }
+    // Only variables defined via function arguments or assignments.
+    if !matches!(
+        binding.kind,
+        BindingKind::Argument | BindingKind::Assignment
+    ) {
+        return None;
+    }
+    // This excludes `global` and `nonlocal` variables.
+    if binding.is_global() || binding.is_nonlocal() {
+        return None;
+    }
+    // Only variables defined in function scopes
+    let scope = &checker.semantic().scopes[binding.scope];
+    if !scope.kind.is_function() {
+        return None;
+    }
+    if !checker.settings.dummy_variable_rgx.is_match(name) {
+        return None;
+    }
+
+    let possible_fix_kind = get_possible_fix_kind(name, scope, checker);
+
+    let mut diagnostics: Vec<Diagnostic> = binding
+        .references
+        .iter()
+        .map(|ref_id| {
+            Diagnostic::new(
+                DummyVariableAccessed {
+                    name: name.to_string(),
+                    fix_kind: possible_fix_kind,
+                },
+                checker.semantic().reference(*ref_id).range(),
+            )
+            .with_parent(binding.start())
+        })
+        .collect();
```

Comment:
@MichaReiser I know you earlier said:

> I think we should change the rule to report the use of the binding rather than the variable declaration because I noticed when reviewing the ecosystem changes, that I always had to search for the variable use to assess whether the violation is correct. Doing so would match clippy's behavior

I think the disadvantages of that are:
- The error here is really in the naming of the variable, not the fact that the variable is used. It's the name of the variable that we're suggesting to the user that they should change, and it's the name that we change for the user as part of the autofix. As such, I find it somewhat surprising that we'd issue diagnostics that point to the uses of the name rather than the name
- If the user uses a dummy variable 5 times, I think this implementation currently means that we'd issue 5 diagnostics regarding the dummy variable. But arguably there's only one mistake here, which is the name of the variable at the point where it's bound.

Issuing a single diagnostic at the point where the variable is bound also has the advantage of being simpler and more performant in terms of our implementation

Reply:
> The error here is really in the naming of the variable, not the fact that the variable is used.

Not entirely. The error is the combination of the two. There are two possible errors:

1. You used the wrong variable
2. The variable is no longer unused

Long term, I think what I'd want is a diagnostic like clippy where the primary diagnostic range is the use:

```
warning: used underscore-prefixed binding                                                                                                                                                                                                                                                                                                                                          
  --> crates\ruff_python_semantic\src\analyze\visibility.rs:38:5
   |
38 |     _decorator_list
   |     ^^^^^^^^^^^^^^^
   |
note: binding is defined here
  --> crates\ruff_python_semantic\src\analyze\visibility.rs:37:20
   |
37 | pub fn is_override(_decorator_list: &[Decorator], semantic: &SemanticModel) -> bool {
   |                    ^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#used_underscore_binding
   = note: `-W clippy::used-underscore-binding` implied by `-W clippy::pedantic`
   = help: to override `-W clippy::pedantic` add `#[allow(clippy::used_underscore_binding)]`

warning: `ruff_python_semantic` (lib) generated 1 warning        
```

and the secondary diagnostic range is where the variable is declared.

The fact that the primary range is the use makes it much easier to *review* the violation because you can start with the use and editors provide the necessary tools to rename the variable from the use. You can also jump to the declaration if you want. The opposite is also true but finding all references tends to be slower and requires a few more clicks (or the use of a keyboard shortcut). 

> If the user uses a dummy variable 5 times, I think this implementation currently means that we'd issue 5 diagnostics regarding the dummy variable. But arguably there's only one mistake here, which is the name of the variable at the point where it's bound.

This can be avoided by only emitting a diagnostic for the first use of the binding.

Reply:
I don't feel too strongly. So happy to go with your version if you do feel strongly. But two additional considerations:

* The rule is named `unused-variable-accessed`. We should rename it if we report it at the declaration side because it doesn't report the *use* (either way, we should rename it to *use*)
* In the long term, showing the first use and the declaration seems easier than showing the declaration and **all** usages.

Reply:
> But two additional considerations:
> 
>     * The rule is named `unused-variable-accessed`. We should rename it if we report it at the declaration side because it doesn't report the _use_ (either way, we should rename it to _use_)
> 
>     * In the long term, showing the first use and the declaration seems easier than showing the declaration and **all** usages.

I agree with both of these. So I think the immediate action point here is to rename the rule (and the file the rule is in). Maybe `variable-name-implies-unused`? @Lokejoke, could you do that renaming?

Reply:
how about `allow(used-dummy-variable)`

Reply:
Sure

Reply:
Yes, I'll do that